### PR TITLE
Finds error #502

### PR DESCRIPTION
**Pull request by enderb-ms**
Checks that Docker is registered in the EventLog service

### CODE REVIEW

**Comment by PatrickLang** [comment]
@StefanScherer any comments? This should fix #502 

Looks good to me

File: windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1 [hunk 1]
```diff
@@ -57,6 +57,9 @@ Describe "Docker is installed" {
             $filesToDump["docker version"] = "dockerversion.txt"            
         } | Should Not Throw
     }
+    It "Docker is registered in the EventLog service" {
+        (Test-Path "HKLM:\SYSTEM\CurrentControlSet\Services\EventLog\Application\docker") -or (Test-Path "HKLM:\SYSTEM\CurrentControlSet\Services\EventLog\Application\com.Docker.Service") | Should Be $true
```
Comment (StefanScherer) [inline windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1]: Tested with Docker 4 Windows Beta 30.1, the `com.Docker.Service` should be `DockerService`

**Comment by StefanScherer** [comment]
@enderb-ms Thank you!

I've tested it in a fresh Windows 10 14393.447 environment with D4W beta 30, it seems it needs some adaption for that.

```
PS C:\> wget https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/837a9d3269f85e8dc45299dbee414d3d
63a577c0/windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1 -usebasicparsin | iex
Checking for common problems
Describing Windows Version and Prerequisites
 [+] Is Windows 10 Anniversary Update or Windows Server 2016 68ms
 [+] Has KB3192366, KB3194496, or later installed if running Windows build 14393 45ms
 [+] Is not a build with blocking issues 32ms
Describing Docker is installed
 [+] A Docker service is installed - 'Docker' or 'com.Docker.Service'  46ms
 [+] Service is running 16ms
 [+] Docker.exe is in path 2.12s
 [-] Docker is registered in the EventLog service 89ms
   Expected: {True}
   But was:  {False}
   61:         (Test-Path "HKLM:\SYSTEM\CurrentControlSet\Services\EventLog\Application\docker") -or (Test-Path "HKLM:\SYSTEM\CurrentControlSet\Services\EventLog\Application\com.Docker.Service") | Should Be $true
   at <ScriptBlock>, <No file>: line 61
Describing User has permissions to use Docker daemon
 [+] docker.exe should not return access denied 52ms
Describing Windows container settings are correct
 [-] Do not have DisableVSmbOplock set to 1 54ms
   Expected: {0}
   But was:  {1}
   75:              $regvalue.VSmbDisableOplocks | Should Be 0
   at <ScriptBlock>, <No file>: line 75
 [+] Do not have zz values set 55ms
Describing The right container base images are installed
 [-] At least one of 'microsoft/windowsservercore' or 'microsoft/nanoserver' should be installed 182ms
   ValidationMetadataException: The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
   ParameterBindingValidationException: Cannot validate argument on parameter 'Property'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
   at <ScriptBlock>, <No file>: line 99
Describing Container network is created
 [+] At least one local container network is available 1.79s
 [+] No more than one NAT network 52ms
 [-] NAT Network's internal prefix does not overlap with external IP' 92ms
   Expected: value was {}, but should not have been the same
   158:           $BinaryIP.Substring(0, $Length) | Should Not Be $BinaryIPSubnet.Substring(0, $Length)               
   at <ScriptBlock>, <No file>: line 158
   at <ScriptBlock>, <No file>: line 153
Showing output from: docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.2-cs2-ws-beta
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: nat null overlay
Swarm: inactive
Security Options:
Kernel Version: 10.0 14393 (14393.447.amd64fre.rs1_release_inmarket.161102-0100)
Operating System: Windows 10 Pro
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 5 GiB
Name: vagrant-10
ID: OTXP:PC4A:GKIF:5RPJ:X26N:ITZ4:WIB5:CUSL:MZOZ:2LKX:LNJQ:MQ6J
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8

Showing output from: docker version
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Thu Oct 27 00:09:21 2016
 OS/Arch:      windows/amd64
 Experimental: true

Server:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64

Showing output from: docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
c66f8c1bc2f2        nat                 nat                 local
f2cf28920692        none                null                local

Warnings & errors from the last 24 hours
Logs saved to C:\\logs_20161129-230800.csv
PS C:\> cat .\logs_20161129-230557.csv
PS C:\>
```

There is a registry key `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\EventLog\Application\DockerService` with `EventMessageFile` = `C:\Windows\Microsoft.NET\Framework64\v4.0.30319\EventLogMessages.dll`

```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\EventLog\Application\DockerService]
"EventMessageFile"=hex(2):43,00,3a,00,5c,00,57,00,69,00,6e,00,64,00,6f,00,77,\
  00,73,00,5c,00,4d,00,69,00,63,00,72,00,6f,00,73,00,6f,00,66,00,74,00,2e,00,\
  4e,00,45,00,54,00,5c,00,46,00,72,00,61,00,6d,00,65,00,77,00,6f,00,72,00,6b,\
  00,36,00,34,00,5c,00,76,00,34,00,2e,00,30,00,2e,00,33,00,30,00,33,00,31,00,\
  39,00,5c,00,45,00,76,00,65,00,6e,00,74,00,4c,00,6f,00,67,00,4d,00,65,00,73,\
  00,73,00,61,00,67,00,65,00,73,00,2e,00,64,00,6c,00,6c,00,00,00
```

It seems that all logs, also the output of dockerd running in D4W are in `C:\ProgramData\Docker\service.txt`

I don't assume that you will support all installation variant, especially D4W, but at least this script should not show errors.

After changing the registry check from `com.Docker.Service` to `DockerService` it works fine.

File: windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1 (hunk 1 above)
Comment (enderb-ms) [inline windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1]: Fixed. Thanks.

**Comment by PatrickLang** [comment]
Hmm, I'm running Docker for Windows version 1.13.0-rc2-beta31 (9123) and it still has `com.docker.service`


`Get-Service *docker* ; docker info` returns

```none
Status   Name               DisplayName
------   ----               -----------
Running  com.docker.service Docker for Windows Service
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.13.0-rc2
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 14965 (14965.1001.amd64fre.rs_prerelease.161104-1700)
Operating System: Windows 10 Enterprise Insider Preview
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 3.999 GiB
Name: DESKTOP-J3HAKNT
ID: SPPO:ZWUE:QPBC:JGBL:LEUM:MS2R:QRGO:7GVV:XRK7:TX56:VM3E:5NBC
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by PatrickLang** [comment]
@dgageot can you comment on the service naming? This is a troubleshooting script that looks for common misconfigurations for Windows containers

**Comment by dgageot** [comment]
@PatrickLang Sorry, I'm not sure I understand what you expect from me. Can you elaborate, please? Thanks a lot!

**Comment by StefanScherer** [comment]
@enderb-ms Thanks for the change. Now works with beta31. @PatrickLang I can confirm that there is a Windows service `com.Docker.Service`, but a registry key `DockerService`for the event log.

```
PS C:\Users\vagrant> wget -usebasicparsing https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/e601d10b21b247e24d1
40a8ecade41c2ce55a272/windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1 | iex
Checking for common problems
Describing Windows Version and Prerequisites
 [+] Is Windows 10 Anniversary Update or Windows Server 2016 626ms
 [+] Has KB3192366, KB3194496, or later installed if running Windows build 14393 244ms
 [+] Is not a build with blocking issues 59ms
Describing Docker is installed
 [+] A Docker service is installed - 'Docker' or 'com.Docker.Service'  107ms
 [+] Service is running 54ms
 [+] Docker.exe is in path 2.1s
 [+] Docker is registered in the EventLog service 91ms
Describing User has permissions to use Docker daemon
 [+] docker.exe should not return access denied 66ms
Describing Windows container settings are correct
 [-] Do not have DisableVSmbOplock set to 1 135ms
   Expected: {0}
   But was:  {1}
   75:              $regvalue.VSmbDisableOplocks | Should Be 0
   at <ScriptBlock>, <No file>: line 75
 [+] Do not have zz values set 169ms
Describing The right container base images are installed
 [+] At least one of 'microsoft/windowsservercore' or 'microsoft/nanoserver' should be installed 182ms
Describing Container network is created
 [+] At least one local container network is available 1.99s
 [+] No more than one NAT network 49ms
 [+] NAT Network's internal prefix does not overlap with external IP' 75ms
Showing output from: docker info
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 17
Server Version: 1.13.0-rc2
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 14393 (14393.447.amd64fre.rs1_release_inmarket.161102-0100)
Operating System: Windows 10 Pro
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 5 GiB
Name: vagrant-10
ID: WEBX:VL3D:QZ6D:A2WI:XTZV:V73I:7VNS:NIM3:HAWS:FSX7:6JXN:RSYA
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

Showing output from: docker version
Client:
 Version:      1.13.0-rc2
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   1f9b3ef
 Built:        Wed Nov 23 17:40:58 2016
 OS/Arch:      windows/amd64

Server:
 Version:             1.13.0-rc2
 API version:         1.25
 Minimum API version: 1.24
 Go version:          go1.7.3
 Git commit:          1f9b3ef
 Built:               Wed Nov 23 17:40:58 2016
 OS/Arch:             windows/amd64
 Experimental:        false

Showing output from: docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
d439d2a0eebf        nat                 nat                 local
29c6398aa12c        none                null                local

Warnings & errors from the last 24 hours
Logs saved to C:\Users\vagrant\logs_20161130-121329.csv
```

**Comment by StefanScherer** [comment]
And I can reproduce my problem. I've created another "Windows 2016 Datacenter - with Containers" machine on Azure and just logged in with RDP and ran the Debug script. The service is not registered to the EventLog.

bildschirmfoto 2016-11-30 um 21 22 13

**Comment by PatrickLang** [comment]
@dgageot - what is the expected service name that's set up by Docker for Windows? I'm seeing `com.docker.service` registered with the service control manager, but `DockerService` in the Windows Application event log. Are these supposed to be different? I'm trying to locate and pull events logged by the Docker daemon managing Windows containers.

**Comment by PatrickLang** [comment]
@StefanScherer thanks for confirming. It looks like Ender's latest change works. We still need to repair the Azure image to create the right registry key automatically. As long as `DockerService` is the expected log name, I think we're good to merge this.

These will be addressed separately:
- https://github.com/docker/for-win/issues/269 - on VSmbDisableOplocks
- Azure VM setup missing the eventlog key (tracked internally to Microsoft)
- Docker OneGet provider scripts missing eventlog key (Ender to file issue)

**Comment by dgageot** [comment]
@PatrickLang What I can tell is that all tests are green on Docker for Windows master branch with `wget -usebasicparsing https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/e601d10b21b247e24d140a8ecade41c2ce55a272/windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1 | iex`

**Pull request merged**